### ISSUE TITLE
How to sync XLA GPU Tensor between torch and torch_xla

### ISSUE THREAD

**Issue by fortianyou** [pytorch/xla#2756]
I'm newly to torch_xla and trying to enable torch_xla in distributed training in PyTorch with multi-node gpu. 
However, it seems torch_xla doesn't support this scenario well，for the following reasons:
1. torch_xla only support single-node multi-processing training by [xmp.spawn](https://pytorch.org/xla/release/1.7/index.html#running-on-multiple-xla-devices-with-multiprocessing)
2. torch_xla GPU aten::Tensor dosen't compatible well with cuda aten::Tensor(since they are difference device)

To workaround the issue, I had try to sync xla tensor gradients and move to cuda aten::Tensor mannually before all-reduce. And something weird found:
1. Each xla tensor sync create a SyncTensorGraph, the compilation slow down very much 
2. Xla aten::ensor conversion to cuda aten::Tensor would actually do copy

## ❓ Questions and Help
1. Is there any function or API that support zero-copy between aten::cuda::Tensor & XLA_GPU aten::tensor?
2. Does each SyncTensor trigger a full-subgraph XLA Compilation?
3. Any best practices or good suggestions to PyTorch multi-node distributed training?

**Comment by byronyi** [comment]
XLA at::Tensor are merely symbolic; they do not represent materialized value unless explicitly requested through sync points (e.g. copy back to cpu).

**Comment by fortianyou** [comment]
> XLA at::Tensor are merely symbolic; they do not represent materialized value unless explicitly requested through sync points (e.g. copy back to cpu).

Yes, XLATensor was used to trace symbolic IR ops. But I think the XLA at::Tensor that I referenced is [XLATensor::tensor_data](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.h#L1245), which should have been materialized after sync.

**Comment by JackCaoG** [comment]
Hi @fortianyou Sorry for the late reply. I don't think we test the multi-node GPU setup very well internally. For you question

**1. Each xla tensor sync create a SyncTensorGraph, the compilation slow down very much.**
This should not happen, this sounds like 
https://github.com/pytorch/xla/blob/ee43111408778968431143471b03f445d48b9e6a/torch_xla/csrc/tensor.h#L174
instead of `SyncTensor`. One thing you could do is to mark_step before accessing the tensor value https://github.com/pytorch/xla/blob/8ff7d3fb2b5cd24d966a5eee38be7fd26e9638fe/torch_xla/csrc/tensor.cpp#L1378
This should materiaze all of the pending IR and you can do whatever you want to do with the tensor without trigger recompiling.

**Xla aten::ensor conversion to cuda aten::Tensor would actually do copy**
This one I am not toally sure, @ailzhang do you have any context?

**Comment by ailzhang** [comment]
> Xla aten::ensor conversion to cuda aten::Tensor would actually do copy

@fortianyou XLA GPU backend is hidden behind XLA backend from pytorch perspective, which means pytorch doesn't have access to the memory address of a XLA GPU backend tensor until it's explicitly materialized, and the materialization comes with a copy. This copy is required since XLA doesn't support storage like CPU or native GPU backend. 
Let us know if this doesn't answer your question!

**Comment by fortianyou** [comment]
> Hi @fortianyou Sorry for the late reply. I don't think we test the multi-node GPU setup very well internally. For you question
> 
> **1. Each xla tensor sync create a SyncTensorGraph, the compilation slow down very much.**
> This should not happen, this sounds like
> https://github.com/pytorch/xla/blob/ee43111408778968431143471b03f445d48b9e6a/torch_xla/csrc/tensor.h#L174
> 
> 
> instead of `SyncTensor`. One thing you could do is to mark_step before accessing the tensor value
> https://github.com/pytorch/xla/blob/8ff7d3fb2b5cd24d966a5eee38be7fd26e9638fe/torch_xla/csrc/tensor.cpp#L1378
> 
> 
> This should materiaze all of the pending IR and you can do whatever you want to do with the tensor without trigger recompiling.
> **Xla aten::ensor conversion to cuda aten::Tensor would actually do copy**
> This one I am not toally sure, @ailzhang do you have any context?

It seems there are multiple why to trigger a Compilation, such as SyncLiveTensors, SyncTensors, ToTensor, GetTensors. What difference between those methods?
